### ISSUE TITLE
permission denied during tmpdir remove after mulled-build

### ISSUE THREAD

**Issue by epruesse** [bioconda/bioconda-utils#172]
TLDR: `with tempfile.TemporaryDirectory()` does not force delete but fails on permission denied.

I've got a recipe (`megahit` bioconda/bioconda-recipes#5877) failing the mulled-test because of a test result file (`final.contigs.fa`) that can't be removed when leaving the `with tempfile.TemporaryDirectory() as d` at the bottom of `pkg_test.py` ([log](https://travis-ci.org/bioconda/bioconda-recipes/jobs/269432161 ), traceback at bottom).

The exception occurs inside of `TemporaryDirectory.cleanup()` in `shutil.rmtree()` due to `permission denied` while trying to unlink a file inside of a write protected directory. (Or perhaps wrong user, as a result of the docker execution).

`shutil.rmtree`, and by extension the `with TemporaryDirectory()`, only do the equivalent of `rm -r` and provide no easy option to do a `rm -rf`. Here's some sample code to reproduce:

```python
import tempfile, os, stat
for perm in 0, stat.S_IREAD|stat.S_IEXEC:
 with tempfile.TemporaryDirectory() as d:
  os.mkdir(os.path.join(d, "tmp"))
  with open(os.path.join(d, "tmp", "file"), "w") as f:
    f.write("test")
    os.chmod(os.path.join(d, "tmp"), perm)
```

Fixing this in a pythonic way is tedious. Bare code:

```python
@contextlib.contextmanager
def TempDir():
  tmpdir = tempfile.TemporaryDirectory()
  yield tmpdir.name

  def onerr(func, path, _):
      if func in (os.listdir, os.open):
          os.chmod(path, stat.S_IWRITE|stat.S_IREAD|stat.S_IEXEC)
          # there is no way to return the listing or fp to shutil.rmtree, so we have to
          # delete the contents of the failed directory
          shutil.rmtree(path, onerror=onerr)
          if func == os.listdir:
              # os.listdir is used in the unsafe (not using fd) path and unlike the safe
              # fd (version) will try to unlink the directory itself after failure. so we
              # give it one:
              os.mkdir(path)
      else: # func in (os.unlink, os.islink, os.open(dir), os.rmdir
          os.chmod(os.path.dirname(path), stat.S_IWRITE|stat.S_IREAD|stat.S_IEXEC)
          func(path)

   shutil.rmtree(tmpdir.name, onerror=onerr)
```

Considering how pretty the above is, I'd go for a KISS solution and use the tried and tested `rm -rf` (which will try it's best to get rid of everything, but never fail). 
```python
@contextlib.contextmanager
def TempDir():
  tmpdir = tempfile.TemporaryDirectory()
  yield tmpdir.name
  sp.run(['rm', '-rf', tmpdir.name])
```

Traceback:
```
Traceback (most recent call last):
  File "/anaconda/bin/bioconda-utils", line 11, in <module>
    load_entry_point('bioconda-utils==0.10.0', 'console_scripts', 'bioconda-utils')()
  File "/anaconda/lib/python3.5/site-packages/bioconda_utils/cli.py", line 537, in main
    argh.dispatch_commands([build, dag, dependent, lint, duplicates, bioconductor_skeleton])
  File "/anaconda/lib/python3.5/site-packages/argh/dispatching.py", line 328, in dispatch_commands
    dispatch(parser, *args, **kwargs)
  File "/anaconda/lib/python3.5/site-packages/argh/dispatching.py", line 174, in dispatch
    for line in lines:
  File "/anaconda/lib/python3.5/site-packages/argh/dispatching.py", line 277, in _execute_command
    for line in result:
  File "/anaconda/lib/python3.5/site-packages/argh/dispatching.py", line 260, in _call
    result = function(*positional, **keywords)
  File "/anaconda/lib/python3.5/site-packages/bioconda_utils/cli.py", line 410, in build
    mulled_upload_target=mulled_upload_target,
  File "/anaconda/lib/python3.5/site-packages/bioconda_utils/build.py", line 352, in build_recipes
    docker_builder=docker_builder,
  File "/anaconda/lib/python3.5/site-packages/bioconda_utils/build.py", line 164, in build
    res = pkg_test.test_package(pkg_path, base_image=base_image)
  File "/anaconda/lib/python3.5/site-packages/bioconda_utils/pkg_test.py", line 147, in test_package
    p = utils.run(cmd, env=env, cwd=d)
  File "/anaconda/lib/python3.5/tempfile.py", line 808, in __exit__
    self.cleanup()
  File "/anaconda/lib/python3.5/tempfile.py", line 812, in cleanup
    _shutil.rmtree(self.name)
  File "/anaconda/lib/python3.5/shutil.py", line 474, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/anaconda/lib/python3.5/shutil.py", line 412, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/anaconda/lib/python3.5/shutil.py", line 432, in _rmtree_safe_fd
    onerror(os.unlink, fullname, sys.exc_info())
  File "/anaconda/lib/python3.5/shutil.py", line 430, in _rmtree_safe_fd
    os.unlink(name, dir_fd=topfd)
PermissionError: [Errno 13] Permission denied: 'final.contigs.fa'
```

**Comment by bgruening** [comment]
I'm wondering if it would be easier to tell mulled/involucro to remove these files. For example here: https://github.com/galaxyproject/galaxy-lib/blob/master/galaxy/tools/deps/mulled/invfile.lua#L123

**Comment by epruesse** [comment]
It sounds cleaner, too. 

(I just have zero experience with mulled so I'm treating it as a black box -- can't read up on every project... ;)
